src/slicer.cpp

```diff
@@ -53,32 +53,68 @@ void SlicerLayer::makeBasicPolygonLoop(const Mesh* mesh, Polygons& open_polyline
     open_polylines.add(poly);
 }
 
+int SlicerLayer::tryFaceNextSegmentIdx(const Mesh* mesh, const SlicerSegment& segment, int face_idx, unsigned int start_segment_idx) const
+{
+    decltype(face_idx_to_segment_idx.begin()) it;
+    auto it_end = face_idx_to_segment_idx.end();
+    it = face_idx_to_segment_idx.find(face_idx);
+    if (it != it_end)
+    {
+        int segment_idx = (*it).second;
+        Point p1 = segments[segment_idx].start;
+        Point diff = segment.end - p1;
+        if (shorterThen(diff, largest_neglected_gap_first_phase))
+        {
+            if (segment_idx == static_cast<int>(start_segment_idx))
+            {
+                return start_segment_idx;
+            }
+            if (segments[segment_idx].addedToPolygon)
+            {
+                return -1;
+            }
+            return segment_idx;
+        }
+    }
+
+    return -1;
+}
+
 int SlicerLayer::getNextSegmentIdx(const Mesh* mesh, const SlicerSegment& segment, unsigned int start_segment_idx)
 {
     int next_segment_idx = -1;
-    const MeshFace& face = mesh->faces[segment.faceIndex];
-    for (unsigned int face_edge_idx = 0; face_edge_idx < 3; face_edge_idx++)
-    { // check segments in connected faces
-        decltype(face_idx_to_segment_idx.begin()) it;
-        if (face.connected_face_index[face_edge_idx] > -1 && (it = face_idx_to_segment_idx.find(face.connected_face_index[face_edge_idx])) != face_idx_to_segment_idx.end())
+
+    const std::vector<uint32_t> &faces_to_try = segment.endOtherFaces;
+    bool segment_ended_at_edge = faces_to_try.empty();
```

Comment:
Note that it could also be empty because the face was disconnected on that vertex.
That doesn't really seem to be a real performance hit, though, because it will return immediately on line 94.